tests/apollo/test_skvbc_checkpoints.py

```diff
@@ -128,23 +126,79 @@ async def test_checkpoint_propagation_after_crashed_replicas_comeback(self, bft_
+
+        n = bft_network.config.n
+        f = bft_network.config.f
+        c = bft_network.config.c
+
+        self.assertEqual(len(bft_network.procs), n,
+                         "Make sure all replicas are up initially.")
+        current_primary = await bft_network.get_current_primary()
+
+        await self._create_checkpoints_while_crashing(bft_network, skvbc, f, c,
+                                                      current_primary, nb_crashes=2)
+
     async def _crash_replicas_without_primary(
             self, bft_network, nb_crashing):
         current_primary = await bft_network.get_current_primary()
 
-        crashed_replicas = set()
+        crash_replicas = bft_network.random_set_of_replicas(nb_crashing, without={current_primary})
 
-        crash_candidates = bft_network.all_replicas(
-            without={current_primary})
-        random.shuffle(crash_candidates)
-        for i in range(nb_crashing):
-            bft_network.stop_replica(crash_candidates[i])
-            crashed_replicas.add(crash_candidates[i])
+        for crash_replica in crash_replicas:
+            bft_network.stop_replica(crash_replica)
 
-        return crashed_replicas
+        return crash_replicas
 
     async def _bring_up_crashed_replicas(
             self, bft_network, crashed_replicas):
 
         for replica in crashed_replicas:
             bft_network.start_replica(replica)
+
+    async def _create_checkpoints_while_crashing(self,
+                                                 bft_network,
+                                                 skvbc, f, c,
+                                                 current_primary,
+                                                 nb_crashes=20):
+        for _ in range(nb_crashes):
+            crashed_replicas = await self._crash_replicas_without_primary(
+                bft_network=bft_network,
+                nb_crashing=f
+            )
+            self.assertFalse(current_primary in crashed_replicas)
+
+            self.assertGreaterEqual(len(bft_network.procs), 2 * f + c + 1,
+                                    "Make sure enough replicas are up to allow a successful checkpoint creation")
+
+            checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=current_primary)
+
+            try:
+                await skvbc.fill_and_wait_for_checkpoint(
+                    initial_nodes=bft_network.all_replicas(without=crashed_replicas),
+                    checkpoint_num=1,
+                    verify_checkpoint_persistency=False,
+                    assert_state_transfer_not_started=False
+                )
+            except trio.TooSlowError:
+                self.fail("Unable to create enough client requests for checkpoints.")
+
+            await bft_network.wait_for_checkpoint(current_primary, expected_checkpoint_num=checkpoint_before + 1)
+
+            await self._bring_up_crashed_replicas(bft_network, crashed_replicas)
+
+            await trio.sleep(30)
```

Comment:
I'm afraid this would significantly slow down the test. Do we really need to wait for 30 seconds here?

Reply:
For 2 rounds (number of contiguous crashes), 5 seconds wait time is suffice, for 4 or more rounds, higher wait time is required (>= 10 seconds).

Reply:
If you'll notice, none of our tests use sleeps. This is because it makes them subject to non-determinism. Any value running on machine that is slower, or has a temporarily descheduled VM/container can cause the test to fail. This is why we have all these `wait_for` conditions. If you ever feel the need to use a sleep, instead you should reach for an existing wait_for or add your own. This allows optimal test runtime, and longer failure timeouts, since the wait_for will retry much more quickly than any sleep, and the failure timeout can be made larger without having to worry about average test runtime, thus avoiding nondeterminism problems with slow machines.

While I don't like to use the word 'never' much, I'd suggest that we should never use a sleep in any test, for any reason, other than maybe testing an implementation of sleep itself :)

Reply:
Sounds reasonable, @andrewjstone. Will update. :)

Reply:
I have updated the test s.t., stale nodes now wait for checkpoint to propagate to them after they are brought up, instead of waiting for an arbitrary amount of time.